### PR TITLE
Add ProcessThenRemoveExpiredPtr(container, process) to GG/Base.h

### PR DESCRIPTION
**Pull request by LGM-Doyle**
ProcessThenRemovedExpirePtr(container, process) run process on every
weak ptr in container and then removes any expired ptrs.

### CODE REVIEW

File: GG/GG/Base.h [hunk 1]
```diff
@@ -66,6 +67,83 @@ class ScopedAssign
...
+void ProcessThenRemoveExpiredPtrs(
+    Container& container,
+    const std::function<void(std::shared_ptr<typename Container::value_type::element_type>&)>& process,
+    decltype(std::declval<Container>().erase(std::declval<typename Container::iterator>()))* = nullptr,
+    decltype(std::declval<Container>().equal_range(
+                 std::declval<
+                 typename std::add_const<
+                 typename std::add_lvalue_reference<
+                 typename Container::key_type>::type>::type>()))* = nullptr
+)
```
Comment (geoffthemedio) [inline GG/GG/Base.h]: ) to end of previous line

File: GG/GG/Base.h [hunk 2]
```diff
@@ -66,6 +67,83 @@ class ScopedAssign
...
+}
+
+// set, unordered_set
+template <typename Container>
+void ProcessThenRemoveExpiredPtrs(
+    Container& container,
+    const std::function<void(std::shared_ptr<typename Container::value_type::element_type>&)>& process,
+    decltype(std::declval<Container>().erase(std::declval<typename Container::iterator>()))* = nullptr,
+    decltype(std::declval<Container>().equal_range(
+                 std::declval<
```
Comment (geoffthemedio) [inline GG/GG/Base.h]: fix indenting

File: GG/GG/Base.h [hunk 3]
```diff
@@ -66,6 +67,83 @@ class ScopedAssign
...
 
+/** Apply \p process to each weak_ptr in \p container and then remove any that are expired.*/
+
+// vector
+template <typename Container>
+void ProcessThenRemoveExpiredPtrs(
+    Container& container,
+    const std::function<void(std::shared_ptr<typename Container::value_type::element_type>&)>& process,
+    typename std::add_pointer<decltype(std::declval<Container>().at(std::declval<typename Container::size_type>()))>::type = nullptr
+)
```
Comment (geoffthemedio) [inline GG/GG/Base.h]: ) to previous line

**Review by LGM-Doyle — COMMENTED** [review]
Restyled.

**Pull request merged**